apps/site/src/pages/account/api-keys-panel.tsx

```diff
@@ -1,142 +1,116 @@
 import { NextSeo } from "next-seo";
-import { FunctionComponent, useEffect, useMemo, useState } from "react";
+import {
+  FunctionComponent,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
 
 import { Button } from "../../components/button";
-import { WarningIcon } from "../../components/icons";
+import { FontAwesomeIcon } from "../../components/icons";
 import { Link } from "../../components/link";
-import { GenerateApiModal } from "../../components/pages/dashboard/generate-api-modal";
-import { Table, TableRows } from "../../components/table";
-import { DateTimeCell } from "../../components/table-cells";
-import { UserFacingApiKeyProperties } from "../../lib/api/model/api-key.model";
+import { PanelSection } from "../../components/pages/account/panel-section";
 import { apiClient } from "../../lib/api-client";
+import { ApiKeysContext } from "./api-keys-panel/api-keys-context";
+import { ApiKeysEmptyState } from "./api-keys-panel/api-keys-empty-state";
+import { ApiKeysList } from "./api-keys-panel/api-keys-list";
+import {
+  ApiKeyProps,
+  ApiKeysContextValue,
+  KeyActionStatus,
+} from "./api-keys-panel/types";
 
 export const ApiKeysPanel: FunctionComponent = () => {
-  const [activeApiKeys, setActiveApiKeys] = useState<
-    UserFacingApiKeyProperties[]
-  >([]);
+  const [apiKeys, setApiKeys] = useState<ApiKeyProps[]>([]);
+  const [keyActionStatus, setKeyActionStatus] = useState<KeyActionStatus>();
+  const [isCreatingNewKey, setIsCreatingNewKey] = useState(false);
+  const [newlyCreatedKeyIds, setNewlyCreatedKeyIds] = useState<string[]>([]);
 
-  const [generateKeyStatus, setGenerateStatus] = useState<{
-    showModal: boolean;
-    keyToRegenerate?: UserFacingApiKeyProperties | undefined;
-  }>({
-    showModal: false,
-  });
+  const fetchAndSetApiKeys = useCallback(async () => {
+    const { data } = await apiClient.getUserApiKeys();
 
-  const fetchAndSetApiKeys = () =>
-    apiClient
-      .getUserApiKeys()
-      .then(({ data }) =>
-        data
-          ? setActiveApiKeys(
-              data.apiKeysMetadata.filter((key) => !key.revokedAt),
-            )
-          : null,
-      );
+    if (data) {
+      setApiKeys(data.apiKeysMetadata.filter((key) => !key.revokedAt));
+    }
+  }, []);
 
   useEffect(() => {
-    /** @todo handle errors and show the user a msg */
     void fetchAndSetApiKeys();
+    /** this useEffect meant to be only run once */
+    // eslint-disable-next-line react-hooks/exhaustive-deps
   }, []);
 
-  const tableRows: TableRows = useMemo(
-    () =>
-      activeApiKeys.map((key) => [
-        key.displayName,
-        key.publicId,
-        key.lastUsedAt ? (
-          <DateTimeCell key="lastUsed" timestamp={key.lastUsedAt} />
-        ) : (
-          "Never"
-        ),
-        <DateTimeCell key="createdAt" timestamp={key.createdAt} />,
-      ]),
-
-    [activeApiKeys],
+  const apiKeysContextValue: ApiKeysContextValue = useMemo(
```

Comment:
I think this context might be doing too much and is too 'low level', which makes the whole thing less readable/easy to reason about.

I reckon it ought to just be a few utility functions, and remove the data that manage change (i.e `apiKeys`, `newlyCreatedKeyIds`) and just pass them to components or move those state pieces down. Do we really need to share the `setApiKeys` function around the place? It might be much clearer if we move to `revokeApiKey` etc functions from the `ApiKeysList` component to here, and remove `apiKeys` and `setApiKeys` from the context. Happy to pair with you on this.

Reply:
All of the components consuming this context are meant to be the parts of the `api-keys-panel`, and not used anywhere else. It was like the `few utility function` approach at the beginning while I'm coding it, but then I decided to put the top-level state values into the context, and update the context easily from children, instead of prop-drilling. It was prop-drilling before as I said, then I didn't liked that API, so refactored it to be like that. 

If you want we can discuss more about which one is better, but I prefer to do it as a follow-up PR/discussion if possible.

Reply:
I don't think you need prop drilling or this massive context that's difficult to reason about. Let's chat offline about this.

Reply:
I think this massive context is making it take >100ms when you try to add a new row, on my 12 core M2 Pro macbook pro. Let's look at improving this. 
![Uploading image.png…]()

Reply:
resolving here, related changes will be implemented at https://github.com/blockprotocol/blockprotocol/pull/1271

Reply:
Will leave this unresolved until that PR is merged

Reply:
Merged that PR